theano/tensor/nnet/abstract_conv.py

```diff
@@ -2026,26 +2104,42 @@ def perform(self, node, inp, out_):
         o, = out_
         mode = self.border_mode
 
-        if not ((isinstance(mode, tuple) and min(mode) >= 0) or
-                mode in ('valid', 'full', 'half')):
+        if isinstance(mode, tuple):
+            if len(mode) != 2:
+                raise ValueError(
+                    'invalid border_mode {} which must be a '
+                    'tuple of length {}'.format(mode, self.convdim))
+            border = ()
+            for m in mode:
+                if isinstance(m, integer_types) and m >= 0:
+                    border += ((m, m),)
+                elif isinstance(m, tuple) and len(m) == 2 and \
+                        min(m) >= 0:
+                    border += ((int(m[0]), int(m[1])),)
+                else:
+                    raise ValueError(
+                        'invalid border mode {}. The tuple can only contain '
+                        'integers or tuples of length 2'.format(mode))
+            mode = border
+        elif mode not in ('valid', 'full', 'half'):
             raise ValueError(
                 'invalid border_mode {}, which must be either '
-                '"valid", "full", "half", an integer or a tuple of'
-                ' integers'.format(mode))
+                '"valid", "full", "half", an integer or a tuple '
+                'of length {}'.format(mode, self.convdim))
 
         if mode == "full":
-            mode = tuple(dil_kernshp[i] - 1 for i in range(self.convdim))
+            mode = tuple((dil_kernshp[i] - 1,) * 2 for i in range(self.convdim))
         elif mode == "half":
-            mode = tuple(dil_kernshp[i] // 2 for i in range(self.convdim))
+            mode = tuple((dil_kernshp[i] // 2,) * 2 for i in range(self.convdim))
```

Comment:
I think you can combine this with the previous block of ifs. Before your change there was a clear separation: first a check, then a conversion. Now there is some conversion, then a check, then some more conversion.